### ISSUE TITLE
Small ergonomic issue with no token

### ISSUE THREAD

**Issue by steveklabnik** [softprops/hubcaps#19]
Due to the polymorphism of `Github::new()`, if you try to pass in just a `None`, you get a very confusing error:

```rust
src/main.rs:10:18: 10:29 error: unable to infer enough type information about `_`; type annotations or generic parameter binding required [E0282]
src/main.rs:10     let github = Github::new(
                                ^~~~~~~~~~~
```

I fixed this by passing `None::<String>`, but that's a bit awkward.

I'm not sure what the _right_ thing is here.

**Comment by softprops** [comment]
Damn, that's not ideal. I'll try to improve that.

**Comment by steveklabnik** [comment]
One option here is to include two different constructors. What do you think about this API?

```rust
 let client = Client::new();
  let github = Github::new(
    "my-cool-user-agent/0.1.0",
    &client
  );

// using client here is the same as passing a None as before.

let github = github.with_auth_token("personal-access-token");

//now you're using the auth token, like `Some(x)` before.
```

I might be able to send a PR.

**Comment by softprops** [comment]
I think what I'm going to do is make a minor ( but breaking ) change the the existing one.  There's actually a [few other way's to authenticate](https://developer.github.com/v3/#authentication) so I think I'll make the current constructor take an instance of some new trait `Authenticator` which has instances for access tokens, basic auth, and none, perhaps with a default impl of none. 

For now I'm working on porting this over to serde since the recently released serde 0.7 now has everything I held off on using it in the first place. After I do that, I do this and publish a new version. Going forward I don't think there should be any more breaking changes like this.

**Comment by softprops** [comment]
started a branch for this over [here](https://github.com/softprops/hubcaps/tree/credentials)
